### PR TITLE
[Trainer] Fix DP loss

### PR DESCRIPTION
**Pull request by SunMarc**
# What does this PR do?

This PR fixes a couple of issues related to DP: 
- hight fp16 norm_grad -> we need to prepare the optimizer, otherwise, we don't unscale the optimizer when doing gradient clipping + grad scaler
- Fix logic when using `num_items_in_batch`. For DP, when `num_items_per_batch is not None`, the value is by default the sum of the tokens across all devices. This is not the case for DDP for example. So, we need to be careful of that. When using `num_items_in_batch`, we weren't applying DP at all. 
- Fix `num_input_tokens_seen` as it was always logged

Fixes https://github.com/huggingface/transformers/pull/40610, https://github.com/huggingface/transformers/issues/37474 and https://github.com/huggingface/transformers/pull/38938

In any case, users should use DDP instead of DP. I might start adding some warning to prompt users to switch

Fixes the following tests: 
`test_gradient_accumulation`, `test_gradient_accumulation_loss_alignment_with_loss_func`, `test_gradient_accumulation_loss_alignment_with_model_loss`

# Reproducer 

Combination of `fp16`, `average_tokens_across_devices` should all give the same loss approximately

```python
from datasets import load_dataset

from transformers import (
    AutoConfig,
    AutoTokenizer,
    DataCollatorForLanguageModeling,
    GPTNeoXForCausalLM,
    TrainingArguments,
    Trainer
)


config = AutoConfig.from_pretrained('EleutherAI/pythia-14m')
model = GPTNeoXForCausalLM(config=config).to('cuda')
tokenizer = AutoTokenizer.from_pretrained('EleutherAI/pythia-14m')
tokenizer.pad_token = tokenizer.eos_token
train_data = load_dataset("wiwu2390/minipile-100k", split="train")

def tokenize_function(sample):
    return tokenizer(sample["text"], truncation=True, max_length=512)

tokenized_dataset = train_data.map(tokenize_function, batched=True, remove_columns=["text"])

data_collator = DataCollatorForLanguageModeling(
    tokenizer=tokenizer, mlm=False
)

training_args = TrainingArguments(
    output_dir="../data/pythia-14m-minipile-100k",
    num_train_epochs=3,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    gradient_accumulation_steps=4,
    logging_steps=1,
    save_steps=100,
    learning_rate=1e-3,
    weight_decay=0.01,
    warmup_steps=100,
    fp16=True,
)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_dataset,
    tokenizer=tokenizer,
    data_collator=data_collator,
)

trainer.train()
```

### CODE REVIEW

**Comment by SunMarc** [comment]
I've incorporated some of your changes [kaixuanliu](https://github.com/kaixuanliu) !

**Review by ArthurZucker — APPROVED** [review]
Wow very nice 🫡 thanks for fixing a whole bunch of them!

**Review by S1ro1 — APPROVED** [review]
Thanks, lgtm!

> In any case, users should use DDP instead of DP. I might start adding some warning to prompt users to switch

Might be a good time to do so, to have this before v5?
